### DISCUSSION TITLE
Transparent Background in the iTwin Viewer (as an iFrame)

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#2789] [Q&A] [answered]
At [Evercam](https://evercam.io/), we're trying to overlay the iTwin Viwer as an iFrame on top of our app to show a comparison between a Live View from our cameras and the iModel of the client.

## Approach 1:

We were successful in making iTwin work as in iFrame in general.
We are interested in using the [Swiping Comparison](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=swiping-viewport-sample&imodel=Exton+Campus) sample approach.

What we need is to make everything in white below... **transparent**. To allow us to see our app and with it, our camera's live view.

![image](https://user-images.githubusercontent.com/17766221/143482979-4e02b8b3-153a-4033-85ff-9770e718dc3d.png)

So far, we tried making the background transparent by using:
`viewPort.displayStyle.backgroundColor = ColorDef.create(0xFFFFFFFF)` But it did not work.
We tried adding `.withTransparency(100)` & `.withAlpha(255)`  but it also did not work.
We tried using a decimal number converted from a hexadecimal color (like 16764303 for example) in
`viewStateProps.displayStyleProps.jsonProperties?.styles?.backgroundColor` but no transparency was detected.
We also tried this (taken from the 2D geometry creation sample).
`ColorDef.fromTbgr(ColorDef.withTransparency(ColorDef.create(ColorByName.cyan).tbgr, 50))`, still nothing.

## Approach 2:

We thought to go the other way around, bringing the Evercam live view image into the app as a background (just a 2D big panel, not a skybox or a cube map).

We could not find any documentation about the ability to create an image (not a decorator if possible) as part of the scene.

### **We would love to stick to approach 1 as it's easier for us to update the iTwin app as an independent entity without having to maintain two sources of data (Our App + iTwin App).**

**Comment by pmconne** [comment, marked as answer]
Approach 1 should work - it just requires some minor tweaks on our side.

**Reply by oussamabonnor1** [reply]
> Approach 1 should work - it just requires some minor tweaks on our side.
Please let me know when the updates are live ?

**Reply by pmconne** [reply]
PR expected today. The feature ought to be included in 3.0, for which a release candidate is expected on Monday.

**Reply by pmconne** [reply]
In progress as #2791.

**Reply by oussamabonnor1** [reply]
Hello! The PR was merged and a version was released 2 days ago but using the above mentioned code snippets still does not work. Is there a different approach that needs to be applied ?

**Reply by pmconne** [reply]
Be specific about what version you are using please. There has been a patch update to 2.19 and a preliminary release candidate for 3.0. This enhancement is only available in 3.0.
Please also check the CSS for the div containing your viewport to ensure its background is at least somewhat transparent - otherwise the browser will not blend it with elements behind it.

**Reply by oussamabonnor1** [reply]
Currently using `^2.19.26` which doesn't have this yet I guess. As for our CSS we configured the iFrame to have a transparent background.
May I ask if the 3.0 version is still in preview and if so, how should we go about using it (simply upgrade our packages or do we need a special type to permission to preview it) ?

**Reply by pmconne** [reply]
3.0 is not officially released. 3.0.0-dev.146 is a release candidate being used to identify any potential issues before official release.
You are welcome to try it out. If you do, please consult the first few sections of [NextVersion.md](https://github.com/iTwin/itwinjs-core/blob/master/docs/changehistory/NextVersion.md). As this is a new major version, upgrading may require more changes to your app than the typical minor version. A codemod tool is provided to automate some of those changes.

EDIT: Although if you are not running your own backend it may be slightly too early to upgrade. @calebmshafer is 3.0 general-purpose backend available yet?

**Reply by oussamabonnor1** [reply]
May I know when the 3.0 version is planned for official release ?

**Reply by pmconne** [reply]
Tentatively scheduled for next week. Exact date will depend on what issues are identified with the release candidate. We will post an announcement here.

**Reply by oussamabonnor1** [reply]
Any updates on this ? If not then we'll go for the 3.0.0-dev.146 as testing material meanwhile.

**Reply by aruniverse** [reply]
We'll have an update before end of day today

**Reply by aruniverse** [reply]
See https://github.com/iTwin/itwinjs-core/discussions/2942

**Reply by oussamabonnor1** [reply]
Hi, we managed to make the V3 app work but using the `viewPort.displayStyle.backgroundColor = ColorDef.create(0x00FFFFFF).withTransparency(255)` approach seems to not work.
Our Iframe has `allowtransparency="true"` and we know it works because when adding
`.iui-body{
  background-color: transparent !important;
}` to the main scss file the result is:

![image](https://user-images.githubusercontent.com/17766221/148204314-c0270b8f-3721-451e-9e2d-519e1fc17004.png)

But after signing-in, it seems there's a persistent background within a Canvas layer of the viewer itself (probably a UI layer).

![image](https://user-images.githubusercontent.com/17766221/148204646-2987e009-63a1-42d3-907a-2cc125f13b1c.png)

By trial & error, we found that disabling a css property as shown below seems to do the trick but it's such a dirty hack, is there a better way ?

![image](https://user-images.githubusercontent.com/17766221/148205602-c216ab52-7c44-4fa1-bf75-c42bfc9bba41.png)

**After some looking, that variable should be in line 34 of node_modules/@itwin/core-react/lib/esm/core-react/colorvariables.scss**

**Reply by pmconne** [reply]
@NancyMcCallB @bsteinbk is there a better way for @oussamabonnor1 to control the transparency of a ViewportComponent (or whatever appui layer is preventing his background from being visible)?

**Reply by bsteinbk** [reply]
@oussamabonnor1 please try adding
`.nz-widgetPanels-panels.uifw-widgetPanels { background-color: transparent !important; }`
to your main scss file to remove the default background-color.

**Reply by pmconne** [reply]
@oussamabonnor1 did @bsteinbk's suggestion solve the problem for you?

**Reply by oussamabonnor1** [reply]
Yes, thank you!
